src/libraries/System.IO.Compression/src/System/IO/Compression/WinZipAesStream.cs

```diff
@@ -0,0 +1,744 @@
+                System.Buffers.Binary.BinaryPrimitives.WriteInt64LittleEndian(destSpan.Slice(i), destVal ^ srcVal);
+                i += 8;
+            }
+
+            // Handle remaining bytes
+            while (i < count)
+            {
+                destSpan[i] ^= srcSpan[i];
+                i++;
+            }
+        }
+
+        private void IncrementCounter()
+        {
+            // WinZip AES treats the entire 16-byte block as a little-endian 128-bit integer
+            for (int i = 0; i < 16; i++)
+            {
+                if (++_counterBlock[i] != 0)
+                    break;
+            }
+        }
+
+        private async Task WriteAuthCodeCoreAsync(bool isAsync, CancellationToken cancellationToken)
+        {
+            if (!_encrypting || _authCodeValidated)
+                return;
+
+            _hmac.TransformFinalBlock(Array.Empty<byte>(), 0, 0);
+            byte[]? authCode = _hmac.Hash;
+
+            if (authCode is not null)
+            {
+                // WinZip AES spec requires only the first 10 bytes of the HMAC
+                if (isAsync)
+                {
+                    await _baseStream.WriteAsync(authCode.AsMemory(0, 10), cancellationToken).ConfigureAwait(false);
+                }
+                else
+                {
+                    _baseStream.Write(authCode, 0, 10);
+                }
+
+                Debug.WriteLine($"Wrote authentication code: {BitConverter.ToString(authCode, 0, 10)}");
+            }
+
+            _authCodeValidated = true;
+        }
+
+        private void WriteAuthCode()
+        {
+            WriteAuthCodeCoreAsync(isAsync: false, CancellationToken.None).GetAwaiter().GetResult();
+        }
+
+        private Task WriteAuthCodeAsync(CancellationToken cancellationToken)
+        {
+            return WriteAuthCodeCoreAsync(isAsync: true, cancellationToken);
+        }
+
+        private async Task<int> ReadCoreShared(Memory<byte> buffer, bool isAsync, CancellationToken cancellationToken)
+        {
+            ObjectDisposedException.ThrowIf(_disposed, this);
+
+            if (_encrypting)
+                throw new NotSupportedException("Stream is in encryption mode.");
+
+            if (!_headerRead)
+                throw new InvalidOperationException("Header must be read before reading data.");
+
+            int bytesToRead = buffer.Length;
+
+            // If we know the total size, ensure we don't read into the HMAC
+            if (_encryptedDataSize > 0)
+            {
+                if (_encryptedDataRemaining <= 0)
+                {
+                    if (!_authCodeValidated)
+                    {
+                        if (isAsync)
+                            await ValidateAuthCodeAsync(cancellationToken).ConfigureAwait(false);
+                        else
+                            ValidateAuthCode();
+                    }
```

Comment:
This block seems to be duplicated multiple times, can we have it only in one place?